File: 1200.最小绝对差.py
#
# @lc app=leetcode.cn id=1200 lang=python3
#
# [1200] 最小绝对差
#

# @lc code=start
class Solution:
    def minimumAbsDifference(self, arr: List[int]) -> List[List[int]]:
        arr.sort()
        if len(arr)<3:
            return [arr]

        min_diff = arr[1] - arr[0]
        res = [arr[0:2]]
        for i in range(2,len(arr)):
            if arr[i] - arr[i - 1] == min_diff:
                res.append([arr[i - 1], arr[i]])
            elif arr[i] - arr[i - 1] < min_diff:
                min_diff = arr[i] - arr[i - 1]
                res = [[arr[i - 1], arr[i]]]
        return res


# @lc code=end

File: 1200.最小绝对差.rs
/*
 * @lc app=leetcode.cn id=1200 lang=rust
 *
 * [1200] 最小绝对差
 */

// @lc code=start
impl Solution {
    pub fn minimum_abs_difference(mut arr: Vec<i32>) -> Vec<Vec<i32>> {
        use std::cmp;
        arr.sort();
        let min = arr[..].windows(2).fold(i32::max_value(), |min, slice| {
            cmp::min(slice[1] - slice[0], min)
        });
        arr.windows(2)
            .filter_map(|slice| {
                if slice[1] - slice[0] == min {
                    Some(vec![slice[0], slice[1]])
                } else {
                    None
                }
            })
            .collect::<Vec<Vec<_>>>()
    }
}
// @lc code=end

